### ISSUE TITLE
doesn't refresh newly installed apps.

### ISSUE THREAD

**Issue by nihal697** [ligi/FAST#113]
Fast doesn't recognize newly installed apps. so to do that I have to clear cache in settings of the app but that also resets sorting of apps and the apps I pinned to the top. Please look into this issue.

**Comment by SebiderSushi** [comment]
This is a known issue. FAST has quite some problems with automatic refreshing of apps that i once attempted to [fix](https://github.com/ligi/FAST/pull/100) in my own fork but i couldn't yet make it compatible with android oreo and the changes also haven't been merged.

In the meantime as a workaround you can use the fact that FAST refreshes on a new start and ends the main activity when entering the settings. So going to settings and back right away will trigger a refresh without the need to clear any data.

Or, if you are familiar with compiling android apps and don't run Android Oreo or newer you could go ahead and compile for yourself with my changes.

**Comment by nihal697** [comment]
Oh, didn't know that. Thanks for replying so fast. 
I have one suggestion also, should I say it here or in the pull request section?

**Comment by SebiderSushi** [comment]
If your suggestion concerns something to be implemented into the pull request then i think you should post it there.
I'm glad to help ^^

**Comment by nihal697** [comment]
Actually I am new here and I don't understand what exactly pull request means. So let me tell ny suggestion here and if it is inappropriate then I will delete it and add it to pull request.
Suggestion is to add a feature which opens apps automatically when the string of letters matches only a single app. I saw it on an open source app the link of whose sources I am providing here :
https://github.com/vackosar/search-based-launcher/
Tell me if I should delete it from here, please. Hope you like the idea.

**Comment by nihal697** [comment]
I am literally sorry that I didn't notice it is already in the settings. Extremely sorry sir. (._.)

**Comment by SebiderSushi** [comment]
No Problem, settings of new apps can be quite confusing sometimes. :D

**Comment by nihal697** [comment]
Certainly it is. Thank you again. ❤️
Will close the issue as you said it is already known.

**Comment by SebiderSushi** [comment]
I think you should leave it open until it is fixed for good.

**Comment by nihal697** [comment]
done.

**Comment by SebiderSushi** [comment]
I've just completed my rework of FASTs app list update mechanisms. The only thing necessary now is to test it and polish it up before release. If you would want to test this it would greatly be appreciated as I can't properly test it under real life conditions since I don't use Android 8.0 or newer on my main device.
If you are familiar with building android apps go ahead and compile my feature branch.
If not, I have uploaded a [testing build](https://github.com/SebiderSushi/FAST/releases/tag/redesign_receiever) on my forked repository that can be installed alongside the official FAST application.

**Comment by nihal697** [comment]
Hey, I would really love to help you in testing the app out, although
unfortunately I don't know how to build apps, so I will go for the testing
build. :D

**Comment by nihal697** [comment]
Fast-debug_redesign_test_v00.apk
The test build is with this name right, just to confirm?

**Comment by SebiderSushi** [comment]
Yes :)

**Comment by nihal697** [comment]
It isn't being installed, i am running on a custom rom based on 8.1

**Comment by nihal697** [comment]
It finally got installed when i did that from mixplorer's inbuilt
installer, strangely android package installer was unable to do this. I
disabled one app just now and it got updated quickly, will report further
if any issues found. Thanks

**Comment by SebiderSushi** [comment]
Very strange. But good that it worked anyway. Thank you very much for putting your trust and time into this and testing!

**Comment by nihal697** [comment]
That's just your generosity sir, oh and I don't trust you at all btw :p
I tried disabling and uninstalling apps and then quickly checking them in
the FAST, it was working fine each and every time so i guess this mechanism
is absolutely working well. :D

**Comment by SebiderSushi** [comment]
But installing an app someone else built sure needs to happen out of at least a little bit of trust.

**Comment by nihal697** [comment]
Okay then, I love you too.

**Comment by nihal697** [comment]
Hello, It's been quite a time and I didn't find any issue in FAST except that most of the times if an app is uninstalled from another launcher or from settings directly, FAST doesn't recognise the uninstallation. Thanks. :D

**Comment by SebiderSushi** [comment]
Okay this should not happen.
Could you give the details specified in the last paragraph of the [release notes](https://github.com/SebiderSushi/FAST/releases/tag/redesign_receiever) of the testing build to give a clearer image?
Also if you might know anything that could be important when reproducing this bug that would be helpful too.

**Comment by nihal697** [comment]
Yep sure
1. 116 apps
2. Icon resolution is 48
3. I can't really tell about the workload as I didn't even notice it. :(
4. Fast isn't set as default.
To reproduce the issue you can set some other launcher as default and uninstall any app either from that launcher or from android settings. FAST won't notice the uninstallation most probably. Even force stopping FAST doesn't work. It isn't the case if it is set as default launcher. 


-------- Original Message --------

**Comment by SebiderSushi** [comment]
Thank You very much for the information. I will look into it when i find the time.

**Comment by nihal697** [comment]
You're welcome and thank you too. 

On June 10, 2019 7:34:43 AM UTC, Sebastian <notifications@github.com> wrote:
>Thank You very much for the information. I will look into it when i
>find the time.
>
>-- 
>You are receiving this because you modified the open/close state.
>Reply to this email directly or view it on GitHub:
>https://github.com/ligi/FAST/issues/113#issuecomment-500321590

-- 
Sent from my Android device with Librem Mail. Please excuse my brevity.

**Comment by SebiderSushi** [comment]
Okay i have been using the testing build for about a week now since i finally have a daily driver phone running Oreo (more precisely LineageOS 15.1).
Having set Trebuchet as the default launcher i could not reproduce your observations, being FAST not recognizing an app being uninstalled from Trebuchet. May i ask which ROM you are using?

The only issue i found to be reproducible on my system seems to be that on Oreo FAST doesn't recognize it automatically if a "component" is enabled or disabled. For example there are Apps that offer to manually specify in their settings to hide or show one of their activities. I also found the error in the code that probably causes this and will fix it right now.

**Comment by SebiderSushi** [comment]
**Correction:** While investigating the issue of FAST not detecting components correctly i managed to reproduce your error: The issue i was investigating turned out to consist of two parts, one of them being the error in the code i already found and the other one being the issue you were observing.

The fixed version can now be downloaded from my [release page](https://github.com/SebiderSushi/FAST/releases) (just pick the latest one).

#### Explanation
&nbsp;It happended to be that if the list of changed packages was too short, then the refresh mechanism that should refresh only said changed packages on a fresh FAST launch (i.e. FAST has been terminated some time before) would be very quick. Due to a mistake i made this would lead to a situation where (i guess) somehow the mechanism would save the refreshed app list to disk while the user interface would have already loaded the previous version from disk and therefore keeps displaying that and then saves that outdated list later when you leave FAST, overwriting the refreshed list.
&nbsp;I probably always (un)installed multiple apps while trying to reproduce your issue. In these cases the refreshing mechanism would take long enough for the user interface to completely initalize. At that point instead of writing the new list to disk, it will be sent directly to the running user interface and so the changes are applied.
&nbsp;My fix simply consists in making sure that the callback resposible for sending the updated list directly to the user interface gets initialized before the updating mechanism is executed.

**Comment by nihal697** [comment]
That's really great to hear, although I didn't understand the technicalities much. Thanks a lot sir, once again! :D

**Comment by nihal697** [comment]
Hello
I know it's been a long time, so how are you?
i was just confused about if you ever pushed the test builds of Fast app to
fdroid/playstore?
thanks.

**Comment by SebiderSushi** [comment]
I pushed a new build [on my release page.](https://github.com/SebiderSushi/FAST/releases)
It's still not up to code but should work. I put together the existing code and fixed some bits here and there.

At some point i stopped using FAST and stopped working on the patch. It was left lingering for about 3 years now and is 5 years old in total. I wouldn't get my hopes up that it will ever get merged at this stage. Right now it is what it is but it should work pretty ok.

**Comment by nihal697** [comment]
yeah, that's totally understandable. Much thanks and appreciation for
pushing another build for just a guy :)
I dont understand the difference between extras and no extras version. Can
you tell me?
Also, which launcher do you use nowadays ?
Thanks again

**Comment by SebiderSushi** [comment]
I'm using Lawnchair. The inability to use app shortcuts or widgets on FAST was always a bummer...

---


The TL;DR is: You likely only care about `forFDroid-withExtras`
If you're curious, build variants are different in these aspects:

The noExtras version is set up to get the app as small as possible. It is what you get when you download the app "FASTER" [(org.ligi.faster on F-Droid)](https://f-droid.org/en/packages/org.ligi.faster).  
&nbsp;The are no translations and no "About" page (opened by tapping the question mark in the top bar in the full release).

Nowadays this build variant isn't really useful as almost no contemporary phone would notice a difference between a 60KB apk and a 90KB apk. But it can still be compiled so here we are.

The Amazon, F-Droid, Play variants are basically different in which link they use when you tap "Share" or where you get sent to when you tap "open in market" (only shown on apps that were installed by google play / amazon app store).

**Comment by nihal697** [comment]
got it. So there is no difference in terms of features or fastness of the
opening. Thanks.
I am also using lawnchair, just wanted to make fast app as default
assistant so that i can open apps from any screen.
Thanks a lot man, you're great.

**Comment by nihal697** [comment]
Hi, hope you are doing well.
I saw that you are developing FAST app again.
I wanted to ask if there is some reason you made it show in recents, bcoz
its kinda pain now.
And a request if you find it worth it, please make FAST bottom oriented.
Regards

**Comment by SebiderSushi** [comment]
I thought it could be useful to have fast in recents when not using it as a launcher but i can revert this change, sure. Can you go into more detail as to why it is a pain? I'd like to grasp this a little better.

As for a bottom-oriented layout, there's issue #66 which has been around since 2015. Again, there's not much going on with this project anymore. My intention is to finally finish my work that i started and make FAST at least usable even if it's still far from polished. The core of my changes still is "make FAST refresh the app list properly for once", the rest is what's required to even compile it and some low hanging fruit that i thought i just might fix along the way as well.

**Comment by nihal697** [comment]
Oh okay , understood.
The thing with recents is that
1) if it is used as default launcher, it shows up in recents. Nobody goes
to the launcher by swiping through the apps in recents menu, but by
pressing the home button. So it takes unnecessary space there.
2) if it is used as an assitant app, I personally find it quicker to open
FAST by long pressing the home, instead of scrolling through recents.
Basically same reason as above.
Thanks a lot for replying and working on the project. :D

**Comment by SebiderSushi** [comment]
1) I realize that your default launcher should never appear on the recents screen, but from my testing i came to the conclusion that android already takes care of this situation. I tested my build on CyanogenMod 7 (android 2.3.7), LineageOS 15.1 (android 8.1) and LineageOS 18.1 (android 11) and all of these do not add FAST to the recents list if it is set as the default launcher. What android version by which vendor are you using?
2) Using FAST as an assistant app is indeed a use case i did not test so far so i just tried it out.

After some tests i came to the conclusion that android does not offer any way to make FAST appear on the recents screen with just the regular launcher entry but not the Assist entry. I reverted the change and posted a [new release](https://github.com/SebiderSushi/FAST/releases).

**Comment by nihal697** [comment]
1. im on android 13, limeage os.
 I had not tested it thoroughly but i did now and it seems that it happens
if you go to FAST settings menu. Then the app is stuck on recents menu
unless it is swiped away manually.

**Comment by SebiderSushi** [comment]
I can't reproduce this issue, but i don't have android 13.
You might have to remove FAST from recents and reboot after the update until it all woks as expected again - but that's a shot in the dark i have no clue what android 13 is doing here.

**Comment by nihal697** [comment]
i was talking about the second last update.
the latest one with the changes reverted is absolutely working.

**Comment by SebiderSushi** [comment]
v6.8-redesign-7 will show up on the recents screen just like any other app would, regardless of whether you navigated to the settings screen or not - that's expected behavior. The only exception is if android removes it because it is set as the default launcher app.

glad to hear that v6.8-redesign-8 works properly again.

**Comment by nihal697** [comment]
yeah, thank you again for doing this much!

**Comment by nihal697** [comment]
Hi Sebastian, hope you doing great.
I was wondering if you still interested in maiantaining fast app tool, if
yes id like to request for some small features.
yeah right im still using your app on android 13 cuz theres no better
replacement, there is one app search app that came out recently and has
pretty many features, looks mice but isnt "fast" and "light".
so please do let me know if you are interested.
Thanks again and hoping you're good healthwise and financially.
Love you too
